### PR TITLE
Feature/vdx 171

### PR DESCRIPTION
**Pull request by sksadjad**
features:
- adds structure for issuer
- some structure fixes for the parent project (further fixes will be added after @hrehman-sphereon branch merges)
- added builders
- added tests

### CODE REVIEW

File: packages/client/package.json [hunk 1]
```diff
@@ -37,13 +38,16 @@
...
+    "npm-run-all": "^4.1.5",
     "open-cli": "^7.0.1",
-    "typescript": "4.9.5",
-    "@types/node": "^18.15.3"
+    "typescript": "4.9.5"
   },
   "engines": {
     "node": ">=16"
   },
+  "engines": {
```
Comment (nklomp) [inline packages/client/package.json:48]: Please remove. There already is an engines with node >= 16 in the package.json

File: packages/issuer/lib/VcIssuer.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+import { v4 as uuidv4 } from 'uuid'
+
+export class VcIssuer {
+  _issuerMetadata: ICredentialIssuerMetadataParametersV1_11
+  _preAuthorizedCode?: string
+  _userPinRequired?: boolean
+  constructor(issuerMetadata: ICredentialIssuerMetadataParametersV1_11, preAuthorizedCode?: string, userPinRequired?: boolean) {
+    this._issuerMetadata = issuerMetadata
+    this._preAuthorizedCode = preAuthorizedCode
```
Comment (nklomp) [inline packages/issuer/lib/VcIssuer.ts:17]: Hmzz, not sure whether something which is request specific should be part of the issuer, which should be able to handle multiple request responses in the future

File: packages/issuer/lib/VcIssuer.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    this._userPinRequired = userPinRequired
+  }
+
+  public getIssuerMetadata() {
+    return this._issuerMetadata
+  }
+
+  public createCredentialOfferDeeplink(): string {
```
Comment (nklomp) [inline packages/issuer/lib/VcIssuer.ts:25]: I rather would see this moved to a seperate function outside of the issuer class

File: packages/issuer/lib/VcIssuerBuilder.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+  public withCredentialsSupported(credentialSupported: ICredentialSupportedV1_11 | ICredentialSupportedV1_11[]): VcIssuerBuilder {
+    if (!Array.isArray(credentialSupported))
+      this.credentialsSupported = this.credentialsSupported ? [...this.credentialsSupported, credentialSupported] : [credentialSupported]
+    else {
+      this.credentialsSupported = this.credentialsSupported ? [...this.credentialsSupported, ...credentialSupported] : credentialSupported
+    }
+    return this
+  }
+
+  public withPreAuthorizedCode(preAuthorizedCode: string): VcIssuerBuilder {
```
Comment (nklomp) [inline packages/issuer/lib/VcIssuerBuilder.ts:57]: Should be done per request, not for the whole issuer

File: packages/issuer/package.json [hunk 5]
```diff
@@ -1,15 +1,21 @@
 {
-  "name": "@sphereon/openid4vci-issuer",
+  "name": "@sphereon/oidvci-issuer",
```
Comment (nklomp) [inline packages/issuer/package.json:2]: Please leave it as openid4vci

**Review by nklomp — CHANGES_REQUESTED** [review]
See some remarks

File: packages/issuer/package.json (hunk 5 above)
Comment (sksadjad) [reply packages/issuer/package.json]: changed to `openid4vci`

File: packages/issuer/lib/VcIssuer.ts (hunk 3 above)
Comment (sksadjad) [reply packages/issuer/lib/VcIssuer.ts]: created another ts file: `CredentialOffer.ts` this is the only function in it right now, but I imagine this class will have a couple more

File: packages/issuer/lib/VcIssuer.ts (hunk 2 above)
Comment (sksadjad) [reply packages/issuer/lib/VcIssuer.ts]: removed, now create deeplink directly receives this as a param

File: packages/issuer/package.json [hunk 6]
```diff
@@ -1,5 +1,5 @@
 {
-  "name": "@sphereon/oidvci-issuer",
+  "name": "@sphereon/oidvci",
```
Comment (nklomp) [inline packages/issuer/package.json]: openid4vci-issuer !

Comment (sksadjad) [reply packages/issuer/package.json]: changed to `openid4vci-issuer`

File: packages/issuer/lib/api.ts [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore
+    this.express.get('/metadata', (request, response) => {
+      return response.send(this._vcIssuer._issuerMetadata)
+    })
+  }
+  private registerTokenRequestEndpoint() {
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore
+    this.express.pose('/token', (request, response) => {
```
Comment (maikel-maas) [inline packages/issuer/lib/api.ts:109]: Do you mean post?

File: packages/issuer/lib/CredentialSupportedV1_11Builder.ts [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+      this.types = this.types ? [...this.types, type] : [type]
+    } else {
+      this.cryptographicBindingMethodsSupported = this.cryptographicBindingMethodsSupported
+        ? [...this.cryptographicBindingMethodsSupported, ...type]
+        : type
+    }
+    return this
+  }
+  withCryptographicBindingMethod(method: string | string[]): CredentialSupportedV1_11Builder {
+    if (!Array.isArray(method)) {
```
Comment (maikel-maas) [inline packages/issuer/lib/CredentialSupportedV1_11Builder.ts:40]: If method is not an array and cryptographicBindingSupported is not present you return method as an array? Isn't it the other way around?

File: packages/issuer/lib/VcIssuerBuilder.ts [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    return this
+  }
+
+  public withBatchCredentialEndpoint(batchCredentialEndpoint: string): VcIssuerBuilder {
+    this.batchCredentialEndpoint = batchCredentialEndpoint
+    return this
+  }
+
+  public withIssuerDisplay(issuerDisplay: IIssuerDisplay | IIssuerDisplay[]): VcIssuerBuilder {
+    if (!Array.isArray(issuerDisplay)) this.issuerDisplay = this.issuerDisplay ? [...this.issuerDisplay, issuerDisplay] : [issuerDisplay]
```
Comment (maikel-maas) [inline packages/issuer/lib/VcIssuerBuilder.ts:34]: If isseurDisplay is not an array and this.issuerDisplay is not present you return issuerDisplay as an array? Isn't it the other way around?

File: packages/issuer/lib/api.ts [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+    // @ts-ignore
+    this.express.get('/credential-offer/:pre_authorized_code', async (request, response) => {
+      const preAuthorizedCode = request.params.pre_authorized_code
+      const id = uuidv4()
+      this.tokenToId.set(preAuthorizedCode, id)
+      return response.send(createCredentialOfferDeeplink(preAuthorizedCode, this._vcIssuer._issuerMetadata))
+    })
+  }
+  /*
+      private timeout(ms: number) {
```
Comment (maikel-maas) [inline packages/issuer/lib/api.ts:132]: Remove commented out code

File: packages/issuer/lib/api.ts [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+    .withIssuerDisplay({
+      name: process.env.issuer_name as string,
+      locale: process.env.issuer_locale as string,
+    })
+    .withCredentialsSupported(credentialsSupported)
+    .build()
+}
+
+export class RestAPI {
+  public express: Express
```
Comment (maikel-maas) [inline packages/issuer/lib/api.ts:55]: I miss database support? Will it be added later?

**Review by maikel-maas — CHANGES_REQUESTED** [review]

File: packages/issuer/lib/api.ts (hunk 7 above)
Comment (sksadjad) [reply packages/issuer/lib/api.ts]: yeah, typo fixed

File: packages/issuer/lib/CredentialSupportedV1_11Builder.ts [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+    } else {
+      this.cryptographicBindingMethodsSupported = this.cryptographicBindingMethodsSupported
+        ? [...this.cryptographicBindingMethodsSupported, ...method]
+        : method
+    }
+    return this
+  }
+
+  withCryptographicSuitesSupported(suit: string | string[]): CredentialSupportedV1_11Builder {
+    if (!Array.isArray(suit)) {
```
[the remark this answers is not in the document]
Comment (sksadjad) [reply packages/issuer/lib/CredentialSupportedV1_11Builder.ts:53]: no, it's array in the spec, so I get a single or a list, and I create an array out of it

File: packages/issuer/lib/CredentialSupportedV1_11Builder.ts (hunk 8 above)
Comment (sksadjad) [reply packages/issuer/lib/CredentialSupportedV1_11Builder.ts:40]: no, it's array in the spec, so I get a single or a list, and I create an array out of it

File: packages/issuer/lib/CredentialSupportedV1_11Builder.ts [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+    return this
+  }
+
+  withId(id: string): CredentialSupportedV1_11Builder {
+    this.id = id
+    return this
+  }
+
+  withTypes(type: string | string[]): CredentialSupportedV1_11Builder {
+    if (!Array.isArray(type)) {
```
[the remark this answers is not in the document]
Comment (sksadjad) [reply packages/issuer/lib/CredentialSupportedV1_11Builder.ts:30]: no, it's array in the spec, so I get a single or a list, and I create an array out of it

File: packages/issuer/lib/CredentialSupportedV1_11Builder.ts [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+  }
+
+  withIssuerCredentialSubjectDisplay(
+    subjectProperty: string,
+    credentialSubjectDisplayNameAndLocale: IIssuerCredentialSubjectDisplayNameAndLocale | IIssuerCredentialSubjectDisplayNameAndLocale[]
+  ): CredentialSupportedV1_11Builder {
+    if (!this.credentialSubject) {
+      this.credentialSubject = {}
+    }
+    if (!Array.isArray(credentialSubjectDisplayNameAndLocale)) {
```
[the remark this answers is not in the document]
Comment (sksadjad) [reply packages/issuer/lib/CredentialSupportedV1_11Builder.ts:77]: no, it's array in the spec, so I get a single or a list, and I create an array out of it

File: packages/issuer/lib/VcIssuerBuilder.ts [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+    else {
+      this.issuerDisplay = this.issuerDisplay ? [...this.issuerDisplay, ...issuerDisplay] : issuerDisplay
+    }
+    return this
+  }
+
+  public withCredentialsSupported(credentialSupported: ICredentialSupportedV1_11 | ICredentialSupportedV1_11[]): VcIssuerBuilder {
+    if (!Array.isArray(credentialSupported))
```
[the remark this answers is not in the document]
Comment (sksadjad) [reply packages/issuer/lib/VcIssuerBuilder.ts:42]: no, it's array in the spec, so I get a single or a list, and I create an array out of it

File: packages/issuer/lib/api.ts (hunk 11 above)
Comment (sksadjad) [reply packages/issuer/lib/api.ts:55]: yeah it will be added later. for now we just have a map

File: packages/issuer/lib/api.ts (hunk 10 above)
Comment (sksadjad) [reply packages/issuer/lib/api.ts]: removed

File: packages/issuer/lib/CredentialSupportedV1_11Builder.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (maikel-maas) [reply packages/issuer/lib/CredentialSupportedV1_11Builder.ts:53]: Ok

File: packages/issuer/lib/CredentialSupportedV1_11Builder.ts [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+    if (!Array.isArray(suit)) {
+      this.cryptographicSuitesSupported = this.cryptographicSuitesSupported ? [...this.cryptographicSuitesSupported, suit] : [suit]
+    } else {
+      this.cryptographicSuitesSupported = this.cryptographicSuitesSupported ? [...this.cryptographicSuitesSupported, ...suit] : suit
+    }
+    return this
+  }
+
+  withCredentialDisplay(credentialDisplay: ICredentialDisplay | ICredentialDisplay[]): CredentialSupportedV1_11Builder {
+    if (!Array.isArray(credentialDisplay)) {
```
[the remark this answers is not in the document]
Comment (maikel-maas) [reply packages/issuer/lib/CredentialSupportedV1_11Builder.ts:62]: Ok!

File: packages/issuer/lib/CredentialSupportedV1_11Builder.ts (hunk 8 above)
Comment (maikel-maas) [reply packages/issuer/lib/CredentialSupportedV1_11Builder.ts:40]: Ok!

File: packages/issuer/lib/CredentialSupportedV1_11Builder.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (maikel-maas) [reply packages/issuer/lib/CredentialSupportedV1_11Builder.ts:30]: Ok!

File: packages/issuer/lib/VcIssuerBuilder.ts (hunk 9 above)
Comment (maikel-maas) [reply packages/issuer/lib/VcIssuerBuilder.ts:34]: Ok!

File: packages/issuer/lib/VcIssuerBuilder.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (maikel-maas) [reply packages/issuer/lib/VcIssuerBuilder.ts:42]: Ok!

File: packages/issuer/lib/api.ts (hunk 11 above)
Comment (maikel-maas) [reply packages/issuer/lib/api.ts:55]: Ok!

**Review by maikel-maas — CHANGES_REQUESTED** [review]
I left small remarks

File: packages/common/lib/types/Generic.types.ts [hunk 17]
```diff
@@ -1,4 +1,189 @@
...
+  client_id: string;
+  code_challenge: string;
+  code_challenge_method: CodeChallengeMethod;
+  redirect_uri: string;
+  scope?: string;
+  authorization_details?: AuthorizationDetails[]
+  wallet_issuer?: string
+  user_hint?: string
+}
+
```
[the remark this answers is not in the document]
Comment (nklomp) [reply packages/common/lib/types/Generic.types.ts:110]: Nope, that was op_state pre v11, so should not be in the abstract,  but in a v11

**Pull request closed without merging**